airqo-platform/AirQo-api
# [PLAT-430] Infrastructure as Code

# [PLAT-430] Infrastructure as Code
**_WHAT DOES THIS PR DO?_**
This PR creates infrastructure using code on a single/multiple cloud service providers. The current implementation is set for GCP. PR does the following:
```
- Configures VPC network
- Configures IP addressing
- Configures firewall rules for various AirQo services. E.g mongodb, k8s nodeport, http, icmp, ssh, etc
- Create instances and configure boot_disk setting (image selection, disk size, etc)
```
Go to the AirQo GCP console to view all the resources up and running

This PR also provides code for setup k8s on your local machine (mimicking the could setup) using `Vagrant` and `Ansible`.

**_WHAT JIRA STORY/TASK IS RELATED TO THIS PR?_**
[PLAT-430]
**_WHAT IS THE LINK TO THE PR BRANCH_**
[Link](https://github.com/airqo-platform/AirQo-api/tree/ft-infra-code)
**_HOW DO I TEST OUT THIS PR?_**
- **[Terraform]:**

Download and install terraform from [HERE](https://www.terraform.io/downloads.html)
Check out the repo `cd AirQo-api/infra/terraform/gcp/staging` for GCP. Same procedure for AWS in the respective directory.
Run the following commands:
```
terraform init
terraform plan
terraform apply
```
To destroy the setup/test infrastructure, run: `terraform destroy`
NB: You need to have a GCP `service account key` with the necessary privilege stored in the `assets` directory with the name `airqo-terraform.json`

- **[Ansible]:**
Development:
To setup up k8s on your local machine, make sure you have Vagrant, Ansible and Virtualbox are installed properly on you machine.
- [Installing Vagrant](https://www.vagrantup.com/docs/installation)
- [Installing Ansible](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html)
- [Installing VirtualBox](https://www.virtualbox.org/wiki/Downloads)

Check out this code and `cd` to `AirQo-api/infra/ansible/development`. Afterwards, run `vagrant up`. If all goes well, you will have a k8s cluster with 1 master and 2 worker nodes up and running. You can run `vagrant ssh master` to access the master node. Then run `kubectl get nodes` to view all active worker nodes.

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_**  
N/A
**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**
N/A
**_ARE THERE ANY RELATED PRs?_**
N/A



[PLAT-430]: https://airqoteam.atlassian.net/browse/PLAT-430
[PLAT-430]: https://airqoteam.atlassian.net/browse/PLAT-430
